ambry-network/src/main/java/com.github.ambry.network/http2/Http2NetworkClient.java

```diff
@@ -17,38 +17,94 @@
 import io.netty.channel.pool.ChannelPool;
 import io.netty.channel.pool.ChannelPoolMap;
+import io.netty.util.AttributeKey;
 import java.net.InetSocketAddress;
+import java.util.Collections;
 import java.util.List;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.GenericFutureListener;
+import java.util.ArrayList;
 import java.util.Set;
+import java.util.concurrent.TimeUnit;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 
 /**
  * A HTTP/2 implementation of {@link NetworkClient}.
- * TODO: More implementation details will be added in next PR.
  */
 public class Http2NetworkClient implements NetworkClient {
   private static final Logger logger = LoggerFactory.getLogger(Http2NetworkClient.class);
   private final EventLoopGroup eventLoopGroup;
   private final ChannelPoolMap<InetSocketAddress, ChannelPool> pools;
+  private final Http2ClientResponseHandler http2ClientResponseHandler;
+  private final Http2ClientMetrics http2ClientMetrics;
+  public static final ResponseInfo WAKEUP_MARKER = new ResponseInfo(null, null, null);
+  static final AttributeKey<RequestInfo> REQUEST_INFO = AttributeKey.newInstance("RequestInfo");
 
   public Http2NetworkClient(Http2ClientMetrics http2ClientMetrics, Http2ClientConfig http2ClientConfig,
       SSLFactory sslFactory) {
+    logger.info("Http2NetworkClient started");
     this.eventLoopGroup = new NioEventLoopGroup(http2ClientConfig.http2NettyEventLoopGroupThreads);
-    this.pools = new Http2ChannelPoolMap(sslFactory, eventLoopGroup, http2ClientConfig);
+    this.pools = new Http2ChannelPoolMap(sslFactory, eventLoopGroup, http2ClientConfig, http2ClientMetrics);
+    this.http2ClientResponseHandler = new Http2ClientResponseHandler(this);
+    this.http2ClientMetrics = http2ClientMetrics;
   }
 
   @Override
   public List<ResponseInfo> sendAndPoll(List<RequestInfo> requestsToSend, Set<Integer> requestsToDrop,
       int pollTimeoutMs) {
-    return null;
+    long startTime = System.currentTimeMillis();
+    List<ResponseInfo> readyResponseInfos = new ArrayList<>();
+    // Send request
+    http2ClientMetrics.http2ClientSendRate.mark(requestsToSend.size());
+    for (RequestInfo requestInfo : requestsToSend) {
+      long streamInitiateTime = System.currentTimeMillis();
+      this.pools.get(InetSocketAddress.createUnresolved(requestInfo.getHost(), requestInfo.getPort().getPort()))
+          .acquire()
+          .addListener((GenericFutureListener<Future<Channel>>) future -> {
+            if (future.isSuccess()) {
+              http2ClientMetrics.http2StreamAcquireTime.update(System.currentTimeMillis() - streamInitiateTime);
+              Channel streamChannel = future.getNow();
+              streamChannel.pipeline().addLast(http2ClientResponseHandler);
+              streamChannel.pipeline().addLast(new AmbrySendToHttp2Adaptor());
+              streamChannel.attr(REQUEST_INFO).set(requestInfo);
+              streamChannel.write(requestInfo.getRequest());
```

Comment:
I think If you do writeAndFlush here, you can remove ctx.flush in AmbrySendToHttp2Adaptor.write method.

There is one more thing to consider, is that when the write method fails, it's not going to be captured by anything. We need to deal with that.

Reply:
It works.

Let me see how to handle write failure.

Reply:
failure handling added.